### ISSUE TITLE
Multiple server scheduling

### ISSUE THREAD

**Issue by cheesegrits** [cj123/assetto-server-manager#521]
I'm playing around with running multiple servers, to support events where I need more than one split.  I haven't even started looking at how to support different pre-booked entry lists (which I suspect can't be done with a shared data folder), but for now, just being able to have two pickup mode servers running the same race would be useful ... for example, I've got about 30 people want to run Cadwell Park next week, which only has 18 slots. 

It'd be nice to just set up and schedule the custom race on my "master", and have it automatically scheduled on the secondary server, so both fire up the same config in unlocked pickup mode, and when the first one fills up people can just join the second one.  Which almost works ... if I restart (or start) the secondary after the race has been set up on the master.  But of course if the secondary is already running, the schedule never gets set up.

I had a quick look at go-fsnotify, and it looks like it wouldn't be too hard to have a "is secondary?" option in the main server config, and if set, and store is set to json, have any secondary servers monitor for changes to the json custom_races store, and basically just re-run rm.InitScheduledRaces() if anything changes.  Only wrinkle would be figuring out how to not trigger if the change was made by the same instance.

Anyway ... I'm having a play with it, but wanted to find out if this was something you'd be interested in merging?  I suspect I'm going to run across other multi-server issues, and am prepared to put whatever work needs doing into solving them.

**Comment by cheesegrits** [comment]
Note that this would also solve one of my "API" issues.  If my custom provisioning, running on a separate web app (our league registration / signup site), which runs on the same physical server, can just write out the JSON race config, and ASM automatically "sees" the changes and sets up scheduling ... that would be peachy keen.

Although I'd still have to solve that issue of detecting if the file was changed by the same instance.

**Comment by cheesegrits** [comment]
I went ahead an implemented something.  Uses fsnotify, but I added "debouncing", so it doesn't get screwed up by a flurry of events ... most file writes wind up firing a bunch of events ("create, chmod, write, create" is a typical sequence from a text editor, or "create, create" from ASM itself), so the debouncing just waits until there hasn't been an event for 500 milliseconds or whatever (kind of like debouncing keyboard input so you don't trigger an expensive function on every key stroke, you wait till they've finished typing).

I'm still working on the best way to know when the file write is coming from itself ... had hoped I could use the mutex maintained by the JSON store, but Go has no way of determining the state of a lock, unfortunately.

But it's working.  I have 3 ASM servers on my league server, and can now use the web UI on one of them to schedule a race, and it get automagically added to the schedule queue on all 3.

**Comment by cj123** [comment]
This sounds like an interesting idea. I'd be interested to see what the code looks like for sure. 

I'm not sure whether fsnotify is a good fit here, it would be better if there was some synchronisation between multiple server manager instances - but I don't know what that would look like. Multi-server is such a vast topic, it's difficult to figure out where to begin with approaching it really.

**Comment by cheesegrits** [comment]
Yeah, huge topic.  I've spent a lot of time thinking about it.  That's part of what my API suggestion was about ... if we had a basic API for getting and setting configs, that'd make multiple server sync easier, if one was considered the 'master' and handled all the UI, all it would need on top of that would be a basic config listing all the servers and their ports in one file.  And would allow for (say) regional splits, with secondaries on different physical servers.

But absent an API, as it is such a vast topic, I've started trying to figure out the simplest approach I can to solve the basic issues that I need fixed.  File monitoring may not be the most elegant way, but it gets the job done, within the existing "best practice" for running multiple servers (the shared JSON folder), without any of the servers needing to be aware of each other's existence.

And it's a very small amount of code, which is functionally black boxed from the user.  If, down the road, we (by which I mean you, lol) come up with a better way of doing it, it'd be trivial to rip this code out and replace it, with no backward compat or config issues.  So it's a kind of short term band aid.

Anyway ... I'll do a PR in the next day or so, once I've done some more testing, and you can decide.  No problem if you don't like it, it was only a couple of hours work, no big loss.
